repo/open.go

```diff
@@ -333,6 +408,54 @@ func addThrottler(ctx context.Context, st blob.Storage) (blob.Storage, throttlin
 	return throttling.NewWrapper(st, throttler), throttler, nil
 }
 
+func upgradeLockMonitor(
+	ctx context.Context,
+	upgradeOwnerID string,
+	st blob.Storage,
+	password string,
+	cacheOpts *content.CachingOptions,
+	lockRefreshInterval time.Duration,
+	lastSync time.Time,
+	now func() time.Time,
+) blob.Storage {
+	var m sync.RWMutex
+
+	cb := func() error {
+		// protected read for lastSync because it will be shared between
+		// parallel storage operations
+		m.RLock()
+		if lastSync.Add(lockRefreshInterval).Before(now()) {
+			m.RUnlock()
+			return nil
+		}
+		m.RUnlock()
+
+		cacheTime, repoConfig, err := ReadAndCacheRepoConfig(ctx, st, password, cacheOpts, lockRefreshInterval)
+		if err != nil {
+			return err
+		}
+
+		// only allow the upgrade owner to perform storage operations
+		if locked, _ := repoConfig.UpgradeLock.IsLocked(now()); locked && upgradeOwnerID != repoConfig.UpgradeLock.OwnerID {
+			return ErrRepositoryUnavailableDueToUpgrageInProgress
+		}
+
+		m.Lock()
+		// prevent backward jumps on lastSync
+		if lastSync.Before(cacheTime) {
+			lastSync = cacheTime
+		}
+		m.Unlock()
+
+		return nil
+	}
+
+	return beforeop.NewWrapper(st,
```

Comment:
this is hard to understand, do we really need all these `cb` variants, can we just have one cb parameter to this wrapper and all it can do is fail?

Reply:
will create a separate wrapper for the uniform callback case

Reply:
created another constructor for `beforeop` that sets up the same wrapper with a uniform callback